routing/endpoint_registry.go

```diff
@@ -0,0 +1,147 @@
+package routing
+
+import (
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type metrics interface {
+	DetectedTime() time.Time
+	InflightRequests() int64
+}
+
+type entry struct {
+	detected         time.Time
+	inflightRequests int64
+}
+
+var _ metrics = &entry{}
+
+func (e *entry) DetectedTime() time.Time {
+	return e.detected
+}
+
+func (e *entry) InflightRequests() int64 {
+	return e.inflightRequests
+}
+
+type EndpointRegistry struct {
+	mu sync.Mutex
+
+	data map[string]*entry
+}
+
+var _ PostProcessor = &EndpointRegistry{}
+
+type RegistryOptions struct {
+}
+
+func normalizeSchemeHost(s, h string) (string, string, error) {
+	// endpoint address cannot contain path, the rest is not case sensitive
+	s, h = strings.ToLower(s), strings.ToLower(h)
+
+	hh, p, err := net.SplitHostPort(h)
+	if err != nil {
+		// what is the actual right way of doing this, considering IPv6 addresses, too?
+		if !strings.Contains(err.Error(), "missing port") {
+			return "", "", err
+		}
+
+		p = ""
+	} else {
+		h = hh
+	}
+
+	switch {
+	case p == "" && s == "http":
+		p = "80"
+	case p == "" && s == "https":
+		p = "443"
+	}
+
+	h = net.JoinHostPort(h, p)
+	return s, h, nil
+}
+
+func (r *EndpointRegistry) Do(routes []*Route) []*Route {
+	const configErrFmt = "Error while processing endpoint fade-in settings: %s, %s, %v."
+	for _, route := range routes {
+		for i := range route.LBEndpoints {
+			epi := &route.LBEndpoints[i]
+			s, h, err := normalizeSchemeHost(epi.Scheme, epi.Host)
+			if err != nil {
+				log.Errorf(configErrFmt, epi.Scheme, epi.Host, err)
+				continue
+			}
+
+			epi.Scheme = s
+			epi.Host = h
```

Comment:
So here your idea is to make endpoint host to be always in "host:port" format. Maybe we should make it earlier during _pre_-processing of routes.

Reply:
Or in the (p *algorithmProvider) Do

Reply:
The fact that confuses me about normalizing scheme and host in the **pre-processing** stage is that `routing.PreProcessor` preprocesses `eskip.Route` instead of `routing.Route`:
```
type PreProcessor interface {
	Do([]*eskip.Route) []*eskip.Route
}
```
instead of 
```
type PostProcessor interface {
	Do([]*Route) []*Route
}
```
I am not sure whether it is possible to persist normalized fields in `eskip.Route` in the right way. But I do agree it would be nice to do this as soon as possible.

Reply:
Additionally, I do remember I tried to do this in the `(p *algorithmProvider) Do` method, but such validation makes some test in filters/fadein fail, for example https://github.com/zalando/skipper/blob/v0.17.32/filters/fadein/fadein_test.go#L261. So I decided not to normalize fields in the `algorithmProvider` postprocessor, but I am doing it just right after this postprocessor, see https://github.com/zalando/skipper/pull/2535/files#diff-9b0b16d06f27b640eb77f6ba0dee84df6a660528e5cccd81839e0330c4b92dd2R1889. It seems to me this place is soon enough for (almost) every other component to see normalized fields and late enough to be easily backward compatible.